ffmpeg-plugin/mcm_audio_tx.c

```diff
@@ -114,29 +117,50 @@ static int mcm_audio_write_header(AVFormatContext* avctx)
 static int mcm_audio_write_packet(AVFormatContext* avctx, AVPacket* pkt)
 {
     McmAudioMuxerContext *s = avctx->priv_data;
+    size_t frame_size = s->tx_handle->frame_size;
     const uint8_t *data = pkt->data;
-    mcm_buffer *buf = NULL;
     int size = pkt->size;
     int len, err = 0;
 
     while (size > 0) {
-        buf = mcm_dequeue_buffer(s->tx_handle, -1, &err);
-        if (buf == NULL) {
-            av_log(avctx, AV_LOG_ERROR, "Dequeue buffer error %d\n", err);
-            return AVERROR(EIO);
+        char *dest;
+
+        if (!s->unsent_mcm_buf) {
+            s->unsent_mcm_buf = mcm_dequeue_buffer(s->tx_handle, -1, &err);
+            if (s->unsent_mcm_buf == NULL) {
+                av_log(avctx, AV_LOG_ERROR, "Initial dequeue buffer error %d\n", err);
+                return AVERROR(EIO);
+            }
         }
 
-        len = FFMIN(buf->len, size);
-        memcpy(buf->data, data, len);
+        len = FFMIN(frame_size, size + s->unsent_len);
+
+        if (len < frame_size) {
+            memcpy((char *)s->unsent_mcm_buf->data + s->unsent_len, data, size);
+            s->unsent_len += size;
+            break;
+        }
```

Comment:
Having `if(){}  else{}` instead of `break;` and `static int mcm_audio_write_trailer(AVFormatContext* avctx)` added part wouldn't be better?

Reply:
Not sure I understand that. Could you elaborate more on your concern?

Reply:
I would move the part:
```c
static int mcm_audio_write_trailer(AVFormatContext* avctx)
{
    McmAudioMuxerContext *s = avctx->priv_data;

    if (s->unsent_mcm_buf) {
        int err;

        /* zero the unused rest of the buffer */
        memset((char *)s->unsent_mcm_buf->data + s->unsent_len, 0,
               s->unsent_mcm_buf->len - s->unsent_len);

        err = mcm_enqueue_buffer(s->tx_handle, s->unsent_mcm_buf);
        if (err) {
            av_log(avctx, AV_LOG_ERROR, "Enqueue buffer error %d\n", err);
            return AVERROR(EIO);
        }
    }
```
back here:
```c
        if (len < frame_size) {
            memcpy((char *)s->unsent_mcm_buf->data + s->unsent_len, data, size);
            s->unsent_len += size;
            // break;
            // <---- HERE
        }
```
and have rest of code wrapped in `else { }` statement.
This way you will not exit this method before putting the buffer back.